File: LeetCode_695_DFS.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        this->grid = grid;
        flag.resize(grid.size());
        for(int i = 0;i < grid.size();i++)
            flag[i].resize(grid[0].size());

        int ans = 0;
        for(int i = 0; i < grid.size(); i++)
        {
            for(int j = 0;j < grid[0].size();j++)
            {
                ans = max(ans, area(i,j));
            }
        }
        return ans;

    }

    int area(int r, int c) {
        if (r < 0 || r >= grid.size() || c < 0 || c >= grid[0].size() || flag[r][c] != 0 || grid[r][c] == 0)
             return 0;
        flag[r][c] = 1;
        return (1 + area(r+1,c) + area(r-1, c) + area(r, c-1) + area(r,c+1));
    }

private:
    vector<vector<int>> grid;
    vector<vector<int>> flag;
};

File: LeetCode_695_DFS.py
class Solution:
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        # flag = [[0 for i in range(len(grid[0]))] for j in range(len(grid))]
        ans = 0
        flag = set()

        def area(r, c):
            if r not in range(len(grid)) or c not in range(len(grid[0])) or (r,c) in flag or not grid[r][c]:
                return 0
            flag.add((r,c))
            return 1 + area(r-1,c) + area(r+1,c) + area(r, c-1) + area(r, c+1)

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                ans = max(ans, area(i,j))
        return ans
